### ISSUE TITLE
Better support onChange on keyDown enter

### ISSUE THREAD

**Issue by charlax** [dmtrKovalenko/material-ui-pickers#948]
**Is your feature request related to a problem? Please describe.**

I'm trying to have a modal display below a static date to allow editing it.

Something like that:

![image](https://user-images.githubusercontent.com/120501/54187853-b862ed00-44ae-11e9-8ab5-97fc26f29ce0.png)

I would like the following behavior for the popover:

* Submit (and hide popover) when pressing enter
* Submit (and hide popover) when choosing a date using the date picker modal
* Hide popover when clicking away (using MUI's `Popover` for that)

I've tried registering for the `onKeyDown` event of `DatePicker`, but it fires *before* `onChange`, which means that at this point I don't have the new date...

```jsx
  const handleKeyDown = event => {
    console.log('keyDown', { event });
    if (event.defaultPrevented) {
      return;
    }

    event.stopPropagation();

    if (event.key === 'Escape') {
      hide();
    }

    if (event.key === 'Enter') {
      submit();
    }
  };
```

I've tried putting the submit logic inside `onChange`, but `onChange` fires when opening the date picker modal, which hides the popover and the containing `DatePicker` element...

**Describe the solution you'd like**

I'm not sure what the solution should look like. Perhaps do not fire `onChange` when clicking on the icon to open the date picker modal?

**Comment by charlax** [comment]
I had some more ideas, but they don't work:

* Use a `setTimeout` to `submit`. This is very hacky and won't work because `submit` captures a reference to the current value (see extract below).
* Find a way to register my `handleKeyDown` after the date picker's textfield... Haven't found a way to do that yet.
* What would be nice would be for `onChange` to include the date **and** the event that triggered the `onChange`. This way I would be able to check if it's for a keypress. `onChange(date, event)`.

```tsx
import * as React from 'react';
import { Formik } from 'formik';
import { makeStyles } from '@material-ui/styles';
import Popover from '@material-ui/core/Popover';

// SIMPLIFIED

const EditInline: React.FC<Props> = (props) => {
  const [anchorEl, setAnchorEl] = React.useState(null);
  const [value, setValue] = React.useState(props.initialValue);
  const isEditing = Boolean(anchorEl);

  const onEdit = ev => {
    setAnchorEl(ev.currentTarget);
  };

  const hide = () => setAnchorEl(null);

  const submit = () => {
    if (value === props.initialValue) {
      console.log('not submitting, value did not change');
      hide();
      return;
    }

    const payload = { [props.name]: value };
    console.log('submitting', { payload });
    client.patch(props.url, payload).then(
      () => {
        hide();
      },
      () => {
        hide();
      }
    );
  };

  const handleKeyDown = event => {
    console.log('keyDown', { event });
    if (event.defaultPrevented) {
      return;
    }

    event.stopPropagation();

    if (event.key === 'Escape') {
      hide();
    }

    if (event.key === 'Enter') {
      submit();
    }
  };

  return (
    <>
      <span onClick={onEdit}>{props.children}</span>
      <Popover
        open={isEditing}
        anchorEl={anchorEl}
        onClose={() => submit()}
      >
        <div>
          <DatePicker
            name="value"
            value={value}
            onChange={value => {
              console.log('onChange', { value });
              setValue(value);
            }}
            autoFocus
            disableOpenOnEnter
            clearable
          />
        </div>
      </Popover>
    </>
  );
};

export default EditInline;
```

**Comment by charlax** [comment]
It would be a fairly simple & backward compatible change actually.

In https://github.com/dmtrKovalenko/material-ui-pickers/blob/develop/lib/src/_shared/DateTextField.tsx

Pass the original event to `commitUpdates(value, ev)` and then forward it to `onChange(newValue, ev)` as well. Would only require changes in `handleBlur` and `handleKeyPress`.

**Comment by charlax** [comment]
What I ended doing (roughly):

```tsx
import dateParse from 'date-fns/parse';
import dateFormat from 'date-fns/format';

...

    if (event.key === 'Enter') {
      if (kind === 'date') {
        newValue = dateFormat(
          dateParse(event.target.value, 'dd-MM-yyyy', new Date()),
          'yyyy-MM-dd'
        );
      }
      submit(newValue);
    }
```

I basically grab the `TextField` content, parse it, format it, and use it to submit to the backend. This is not a great solution because it basically re-implements the parsing and formatting part of `DatePicker`...

**Comment by dmtrKovalenko** [comment]
Closed because of #862

**Comment by charlax** [comment]
Hey @dmtrKovalenko what will be the plan here? #862 does not give a lot of details...
